design-proposals/cluster-orch-integrating-k8s-into-emt.md

```diff
@@ -0,0 +1,253 @@
+# Design Proposal: Integrating Kubernetes into Edge Microvisor Toolkit
+
+Author(s): Hyunsun Moon
+
+Last updated: 5/2/25
+
+## Abstract
+
+The Edge Microvisor Toolkit (EMT) is an operating system specifically designed for hosting Kubernetes, streamlining traditional general-purpose operating systems by including only the essential components needed to run container-based applications. Our experience from previous releases has demonstrated that EMT's design principles—image-based deployment and immutable root filesystem—enhance the reliability and consistency of cluster creation compared to even well-maintained general-purpose operating systems, such as Ubuntu.
+
+This proposal aims to extend these proven approaches to Kubernetes lifecycle management by integrating Kubernetes directly into EMT. The integration involves two key changes: incorporating Kubernetes into the EMT build process and embedding it within the image. While this does not mean that EMT machines will come with Kubernetes pre-installed and running, users will still need to configure and initialize Kubernetes. Given that turning into a Kubernetes node is the primary and only use case for EMT machines, a new option will be introduced during the host registration step to allow for the automatic installation of Kubernetes with pre-configured default settings. This addition will streamline the process, making edge devices ready for use more efficiently.
+
+With this design change, we anticipate reduced and consistent cluster creation times, enhanced reliability by eliminating third-party dependencies, improved security through faster application of CVEs, and an improved user experience by providing a zero-touch option from device onboarding to a cluster-ready state.
```

Comment:
the efficency is debatable imo. What we are really doing is reduing the perception of the touches done by the user. Everything is done in oneshot but if we probably count the touches is still the same. Time to deploy improved? I am not sure because if all the components are not ready (namely caddy) what's going to happen to the cluster installation?

Have we done any PoC? Otherwise we may need to look at the instance_status_details to trigger the cluster creation

Reply:
the only thing we are saving is the time to download the installer and the binary. Maybe we can separate download and install commands in ClusterAgent: the agent does the download of the binary while there are some components not ready yet and the installation command comes later when the CM see that all the EN components are happy

Reply:
We're trying to reduce actual number of touches not just perception by providing an option to automatically bootstrap K8s after the provisioning. If you left this comment before you read automatic cluster creation section yet, please resolve.

Also, "time to deploy" should be improved by embedding K3s into EMT. Everything needed to bootstrap K3s should be already part of the image. Caddy is not required anymore. Cluster-agent's job is to download certificates generated by CAPI and some additional custom configuration provided by user, which doesn't require any external assets, and run bootstrap command with them.

Reply:
ok - but please verify that cluster creation can proceed without any problem if all the components are not ready. Or at least keep in mind

Reply:
We'll use K3s's airgap deployment feature for that, see "Skip download K3s for EMT" section. So, pretty sure it works.